File: frontend/src/sets/repository.ts
import type { SetItem, SetItemType, SetRecord, Song, WorkspaceDb } from '../db/schema';
import { uuidv7 } from '../db/uuid';
import type { SyncEngine } from '../sync/engine';
import { byRank, rankBetween, rankForMove } from './rank';

/**
 * Sets: the running order for one service or gig.
 *
 * Every write here is local-first like the rest of the app. Ordering uses fractional ranks, so
 * adding an item touches one row and reordering touches exactly the row that moved.
 */

export interface SetInput {
  name: string;
  scheduled_for?: string | null;
  venue?: string | null;
  notes?: string | null;
  assigned_members?: string[];
}

/** The window in which a set pins itself for offline use, business rule 5. */
export const AUTO_PIN_DAYS = 14;

export class Sets {
  constructor(
    private readonly db: WorkspaceDb,
    private readonly engine: SyncEngine,
  ) {}

  async create(input: SetInput): Promise<string> {
    const id = uuidv7();

    await this.engine.record('sets', id, 'upsert', {
      name: input.name.trim(),
      scheduled_for: blank(input.scheduled_for),
      venue: blank(input.venue),
      notes: blank(input.notes),
      assigned_members: JSON.stringify(input.assigned_members ?? []),
      pinned: 0,
    });

    return id;
  }

  async update(id: string, changes: Record<string, unknown>): Promise<void> {
    await this.engine.record('sets', id, 'upsert', changes);
  }

  async remove(id: string): Promise<void> {
    await this.engine.record('sets', id, 'delete', {});
  }

  /** Business rule 6: items and overrides come along, the date does not. */
  async duplicate(id: string): Promise<string | null> {
    const original = await this.db.sets.get(id);

    if (original === undefined) {
      return null;
    }

    const copy = await this.create({
      name: `${original.name} (copy)`,
      venue: original.venue,
      notes: original.notes,
      assigned_members: parseList(original.assigned_members),
    });

    for (const item of await this.items(id)) {
      await this.engine.record('set_items', uuidv7(), 'upsert', {
        set_id: copy,
        rank: item.rank,
        song_id: item.song_id,
        item_type: item.item_type,
        content: item.content,
        title_snapshot: item.title_snapshot,
        key_override: item.key_override,
        capo_override: item.capo_override,
        arrangement_id: item.arrangement_id,
        sheet_part_override: item.sheet_part_override,
        sections: item.sections,
        note: item.note,
      });
    }

    return copy;
  }

  async items(setId: string): Promise<SetItem[]> {
    return byRank(
      await this.db.set_items
        .where('set_id').equals(setId)
        .filter((row) => row.deleted_at === null)
        .toArray(),
    );
  }

  /** Appends songs in the order they were selected, each with the title it had at the time. */
  async addSongs(setId: string, songs: Song[]): Promise<void> {
    let last = (await this.items(setId)).at(-1)?.rank ?? null;

    for (const song of songs) {
      last = rankBetween(last, null);

      await this.engine.record('set_items', uuidv7(), 'upsert', {
        set_id: setId,
        rank: last,
        song_id: song.id,
        title_snapshot: song.title,
      });
    }
  }

  async addItem(setId: string, type: SetItemType, content: string): Promise<void> {
    const last = (await this.items(setId)).at(-1)?.rank ?? null;

    await this.engine.record('set_items', uuidv7(), 'upsert', {
      set_id: setId,
      rank: rankBetween(last, null),
      item_type: type,
      content,
    });
  }

  async move(setId: string, from: number, to: number): Promise<void> {
    const items = await this.items(setId);
    const rank = rankForMove(items.map((item) => item.rank), from, to);

    if (rank === null) {
      return;
    }

    await this.engine.record('set_items', items[from]!.id, 'upsert', { rank });
  }

  async updateItem(id: string, changes: Record<string, unknown>): Promise<void> {
    await this.engine.record('set_items', id, 'upsert', changes);
  }

  async removeItem(id: string): Promise<void> {
    await this.engine.record('set_items', id, 'delete', {});
  }
}

function blank(value: string | null | undefined): string | null {
  return value == null || value.trim() === '' ? null : value.trim();
}

export function parseList(json: string | null): string[] {
  if (json === null) {
    return [];
  }

  try {
    const parsed = JSON.parse(json) as unknown;

    return Array.isArray(parsed) ? parsed.filter((item): item is string => typeof item === 'string') : [];
  } catch {
    return [];
  }
}

/**
 * Business rule 5: a set is kept offline when it is explicitly pinned, or when its date falls
 * inside the window — including today, and including a set that has just happened, because a
 * band still reads last night's set on the way home.
 */
export function isAutoPinned(set: Pick<SetRecord, 'pinned' | 'scheduled_for'>, now = new Date()): boolean {
  if (set.pinned === 1) {
    return true;
  }

  if (set.scheduled_for === null) {
    return false;
  }

  const when = Date.parse(set.scheduled_for);

  if (Number.isNaN(when)) {
    return false;
  }

  // Whole days, not hours: a set is "today" all day, and yesterday's set is one day old at
  // breakfast as well as at midnight.
  const days = Math.round((midnight(when) - midnight(now.getTime())) / 86_400_000);

  return days <= AUTO_PIN_DAYS && days >= -1;
}

function midnight(instant: number): number {
  return Math.floor(instant / 86_400_000) * 86_400_000;
}

export const ITEM_TYPES: { value: SetItemType; label: string }[] = [
  { value: 'announcement', label: 'Announcement' },
  { value: 'scripture', label: 'Scripture' },
  { value: 'prayer', label: 'Prayer' },
  { value: 'video', label: 'Video cue' },
  { value: 'blank', label: 'Blank / logo' },
  { value: 'text', label: 'Free text' },
];

File: crates/web/src/sets/repository.rs
//! Reading and writing sets.
//!
//! Every write here is local-first like the rest of the app. Ordering uses fractional ranks, so
//! adding an item touches one row and reordering touches exactly the row that moved — which is
//! what lets two people reorder the same set offline and have both moves survive.

use aurum_core::sets::rank::{rank_between, rank_for_move};
use serde_json::{Map, Value, json};

use crate::db::records::{Set as SetRecord, SetItem, Song, alive};
use crate::db::{Database, DbError};
use crate::sync::SyncEngine;

/// What a set is created or renamed with. The date is what drives the 14-day auto-pin, so it is
/// a first-class field rather than another note.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct SetInput {
    pub name: String,
    pub scheduled_for: Option<String>,
    pub venue: Option<String>,
    pub notes: Option<String>,
    pub assigned_members: Vec<String>,
}

/// The kinds of item a set can hold besides a song. Each one is a thing that happens in a
/// service and has to sit in the running order at the right moment.
pub const ITEM_TYPES: [(&str, &str); 6] = [
    ("announcement", "Announcement"),
    ("scripture", "Scripture"),
    ("prayer", "Prayer"),
    ("video", "Video cue"),
    ("blank", "Blank / logo"),
    ("text", "Free text"),
];

#[derive(Clone)]
pub struct Sets {
    db: Database,
    engine: SyncEngine,
}

fn blank(value: Option<&str>) -> Option<String> {
    value
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_owned)
}

fn object(value: Value) -> Map<String, Value> {
    value.as_object().cloned().unwrap_or_default()
}

impl Sets {
    pub fn new(db: Database, engine: SyncEngine) -> Sets {
        Sets { db, engine }
    }

    pub async fn create(&self, input: &SetInput) -> Result<String, DbError> {
        let id = crate::new_id();

        self.engine
            .record(
                "sets",
                &id,
                "upsert",
                object(json!({
                    "name": input.name.trim(),
                    "scheduled_for": blank(input.scheduled_for.as_deref()),
                    "venue": blank(input.venue.as_deref()),
                    "notes": blank(input.notes.as_deref()),
                    "assigned_members": serde_json::to_string(&input.assigned_members)
                        .unwrap_or_else(|_| "[]".to_owned()),
                    "pinned": 0,
                })),
            )
            .await?;

        Ok(id)
    }

    pub async fn update(&self, id: &str, changes: Map<String, Value>) -> Result<(), DbError> {
        self.engine.record("sets", id, "upsert", changes).await
    }

    pub async fn remove(&self, id: &str) -> Result<(), DbError> {
        self.engine.record("sets", id, "delete", Map::new()).await
    }

    /// Business rule 6: items and overrides come along, the date does not. A copied set is next
    /// month's service, and dating it today would pin it and put it on the wrong Sunday.
    pub async fn duplicate(&self, id: &str) -> Result<Option<String>, DbError> {
        let Some(original) = self.db.get::<SetRecord>("sets", id).await? else {
            return Ok(None);
        };

        let copy = self
            .create(&SetInput {
                name: format!("{} (copy)", original.name),
                scheduled_for: None,
                venue: original.venue,
                notes: original.notes,
                assigned_members: parse_list(original.assigned_members.as_deref()),
            })
            .await?;

        for item in self.items(id).await? {
            self.engine
                .record(
                    "set_items",
                    &crate::new_id(),
                    "upsert",
                    object(json!({
                        "set_id": copy,
                        "rank": item.rank,
                        "song_id": item.song_id,
                        "item_type": item.item_type,
                        "content": item.content,
                        "title_snapshot": item.title_snapshot,
                        "key_override": item.key_override,
                        "capo_override": item.capo_override,
                        "arrangement_id": item.arrangement_id,
                        "sheet_part_override": item.sheet_part_override,
                        "sections": item.sections,
                        "note": item.note,
                    })),
                )
                .await?;
        }

        Ok(Some(copy))
    }

    /// A set's items, in the order they are played.
    pub async fn items(&self, set_id: &str) -> Result<Vec<SetItem>, DbError> {
        let mut items = alive(
            self.db
                .by_index::<SetItem>("set_items", "set_id", &set_id.into())
                .await?,
        );

        // Ranks are strings compared as strings; that is the whole point of them.
        items.sort_by(|left, right| left.rank.cmp(&right.rank));

        Ok(items)
    }

    /// Appends songs in the order they were chosen, each with the title it had at the time —
    /// so a set stays readable after a song is renamed or deleted.
    pub async fn add_songs(&self, set_id: &str, songs: &[Song]) -> Result<(), DbError> {
        let mut last = self
            .items(set_id)
            .await?
            .last()
            .map(|item| item.rank.clone());

        for song in songs {
            let Ok(rank) = rank_between(last.as_deref(), None) else {
                return Ok(());
            };

            self.engine
                .record(
                    "set_items",
                    &crate::new_id(),
                    "upsert",
                    object(json!({
                        "set_id": set_id,
                        "rank": rank,
                        "song_id": song.id,
                        "title_snapshot": song.title,
                    })),
                )
                .await?;

            last = Some(rank);
        }

        Ok(())
    }

    pub async fn add_item(
        &self,
        set_id: &str,
        item_type: &str,
        content: &str,
    ) -> Result<(), DbError> {
        let last = self
            .items(set_id)
            .await?
            .last()
            .map(|item| item.rank.clone());

        let Ok(rank) = rank_between(last.as_deref(), None) else {
            return Ok(());
        };

        self.engine
            .record(
                "set_items",
                &crate::new_id(),
                "upsert",
                object(json!({
                    "set_id": set_id,
                    "rank": rank,
                    "item_type": item_type,
                    "content": content,
                })),
            )
            .await
    }

    /// One row moves, and only one. Two people reordering the same set offline both keep their
    /// move, because neither touched the other's row.
    pub async fn move_item(&self, set_id: &str, from: usize, to: usize) -> Result<(), DbError> {
        let items = self.items(set_id).await?;
        let ranks: Vec<String> = items.iter().map(|item| item.rank.clone()).collect();

        let (Ok(Some(rank)), Some(moved)) = (rank_for_move(&ranks, from, to), items.get(from))
        else {
            return Ok(());
        };

        self.engine
            .record(
                "set_items",
                &moved.id,
                "upsert",
                object(json!({ "rank": rank })),
            )
            .await
    }

    pub async fn update_item(&self, id: &str, changes: Map<String, Value>) -> Result<(), DbError> {
        self.engine.record("set_items", id, "upsert", changes).await
    }

    pub async fn remove_item(&self, id: &str) -> Result<(), DbError> {
        self.engine
            .record("set_items", id, "delete", Map::new())
            .await
    }
}

/// A JSON array column read as the list it holds. Anything unreadable is an empty list: a
/// member list nobody can parse must not stop a set opening.
pub fn parse_list(json: Option<&str>) -> Vec<String> {
    serde_json::from_str::<Vec<String>>(json.unwrap_or("")).unwrap_or_default()
}
